src/replication/replication_manager.cpp

```diff
@@ -0,0 +1,290 @@
+  std::ifstream hosts_file(replication_hosts_path);
+  if (!hosts_file.is_open()) {
+    throw REPLICATION_EXCEPTION(fmt::format("Unable to open file: {}", replication_hosts_path));
+  }
+  std::string line;
+  std::string replica_name;
+  std::string replica_hostname;
+  uint16_t replica_port;
+  for (int ctr = 0; std::getline(hosts_file, line); ctr = (ctr + 1) % 4) {
+    switch (ctr) {
+      case 0:
+        // Ignored line.
+        break;
+      case 1:
+        replica_name = line;
+        break;
+      case 2:
+        replica_hostname = line;
+        break;
+      case 3:
+        replica_port = std::stoi(line);
+        // All information parsed.
+        if (identity_ == replica_name) {
+          // For our specific identity, check that the port is right.
+          NOISEPAGE_ASSERT(replica_port == port_, "Mismatch of identity/port combo in replica config.");
+        } else {
+          // Connect to the replica.
+          ReplicaConnect(replica_name, replica_hostname, replica_port);
+        }
+        break;
+      default:
+        NOISEPAGE_ASSERT(false, "Impossible.");
+        break;
+    }
+  }
+  hosts_file.close();
+}
+
+void ReplicationManager::ReplicaConnect(const std::string &replica_name, const std::string &hostname, uint16_t port) {
+  replicas_.try_emplace(replica_name, messenger_, replica_name, hostname, port);
+}
+
+void ReplicationManager::ReplicaSend(const std::string &replica_name, const ReplicationManager::MessageType type,
+                                     const std::string &msg, bool block) {
+  if (!replication_enabled_) {
+    REPLICATION_LOG_WARN(fmt::format("Skipping send -> {} as replication is disabled."));
+    return;
+  }
+
+  REPLICATION_LOG_TRACE(fmt::format("Send -> {} (type {} block {}): msg size {}", replica_name,
+                                    static_cast<uint8_t>(type), block, msg.size()));
+  bool completed = false;
+  try {
+    messenger_->SendMessage(
+        GetReplicaConnection(replica_name), msg,
+        [this, block, &completed](common::ManagedPointer<messenger::Messenger> messenger,
+                                  const messenger::ZmqMessage &msg) {
+          if (block) {
+            // If this isn't a blocking send, then completed will fall out of scope.
+            completed = true;
+            blocking_send_cvar_.notify_all();
+          }
+        },
+        static_cast<uint64_t>(type));
+
+    if (block) {
+      std::unique_lock<std::mutex> lock(blocking_send_mutex_);
+      // If the caller requested to block until the operation was completed, the thread waits.
+      if (!blocking_send_cvar_.wait_for(lock, REPLICATION_MAX_BLOCKING_WAIT_TIME, [&completed] { return completed; })) {
+        // TODO(WAN): Additionally, this is hackily a messenger exception so that it gets handled by the catch.
+        throw MESSENGER_EXCEPTION("TODO(WAN): Handle a replica dying in synchronous replication.");
+      }
+      lock.unlock();
```

Comment:
I thought that you didn't need to call this for `unique_lock`

Reply:
you are correct